disnake/ui/item.py

```diff
@@ -160,27 +161,27 @@ async def callback(self, interaction: MessageInteraction[ClientT], /) -> None:
         pass
 
 
-I_co = TypeVar("I_co", bound=Item, covariant=True)
+SelfViewT = TypeVar("SelfViewT", bound="Optional[View]")
 
 
-# while the decorators don't actually return a descriptor that matches this protocol,
+# While the decorators don't actually return a descriptor that matches this protocol,
 # this protocol ensures that type checkers don't complain about statements like `self.button.disabled = True`,
-# which work as `View.__init__` replaces the handler with the item
-class DecoratedItem(Protocol[I_co]):
+# which work as `View.__init__` replaces the handler with the item.
+class DecoratedItem(Protocol[I]):
     @overload
-    def __get__(self, obj: None, objtype: Any) -> ItemCallbackType:
+    def __get__(self, obj: None, objtype: Type[SelfViewT]) -> ItemCallbackType[SelfViewT, I]:
         ...
 
     @overload
-    def __get__(self, obj: Any, objtype: Any) -> I_co:
+    def __get__(self, obj: Any, objtype: Any) -> I:
         ...
 
 
 T_co = TypeVar("T_co", covariant=True)
 P = ParamSpec("P")
 
 
-class Object(Protocol[T_co, P]):
+class ItemShape(Protocol[T_co, P]):
```

Comment:
I suggest we replace this Protocol with simply `Callable[P, T_co]`; I didn't realize at the time that `Callable` fulfills the criteria for capturing init args and return type :)

Reply:
Right, I suppose that works too; it would widen the possible input types though, since it allows arbitrary callables and isn't necessarily limited to types. Something like `Callable[P, T_co] & Type[T_co]` would be cool, if we had intersection types.

Reply:
> it would widen the possible input types
I'm aware, however I don't see a reason to restrictict the accepted types. ~~The runtime already supports that~~ if not for the issubclass checks, this would work out of the box.

Reply:
I've gone and loosened the `cls` type requirement in https://github.com/DisnakeDev/disnake/pull/1190/commits/42a5edb3842bc33456ad8344cfbf92eb5f6f49f3, and removed the runtime check in https://github.com/DisnakeDev/disnake/pull/1190/commits/6a4aafbe928d97c140694598ceca9130a13bf223.

While passing a function to a parameter named `cls` doesn't really make sense semantically, you're right that there also isn't really a reason to artificially restrict this as much. If someone is using custom view decorators, I would assume they already know what they're doing :>

That said, I'm not quite sure if/how this should be documented without introducing more confusion, or if we should leave the documentation as `cls: Type[...]` for now.-

Reply:
See, this is where `cls` being pos-only would come useful :kek:
With the new semantics, `factory` would be a better name, and I think at least the docs could be updated to treat it as so.

Reply:
Finally got around to doing this - i've updated the `cls` parameter description in https://github.com/DisnakeDev/disnake/pull/1190/commits/e9a6b75993ec9210be577437c397662a69d7eea4, to make it clear that it now accepts any (matching) callable. Renaming it to something more semantically fitting will have to wait, though.